### PR TITLE
Add VK_KHR_synchronization2 example

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a "new" example that shows usage of the new synchronization features introduced with `VK_KHR_synchronization2`. The sample is based on the compute shader N-Body sample, and allows an easy comparison between the old and new synchronization api.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

## Notes
- I haven't tested with Validation layers yet as the accompanying PR has not yet been merged (https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2554)
- Tested on an NVIDIA RTX 2060 on Windows 10
- This PR also updates Volk and the Vulkan headers to 1.2.170

### CODE REVIEW

**Comment by jeremyg-lunarg** [comment]
Sascha, FYI there will be several PRs after https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2554 .  All of the changes we plan on merging are in the jeremyg-synchronization2 branch on github, but expect force rebases  on that branch...

**Comment by marty-johnson59** [comment]
Hi Sascha, I think we're getting close on this one.  I've added a bunch of reviewers. Can you resolve the conflicts here so we can merge once we get the approvals?  Thx

**Comment by SaschaWillems** [comment]
Sure, just merged with master from upstream to resolve the conflicts.

**Comment by marty-johnson59** [comment]
@AlejandroCosin, @aurolou, @carlosadcq, @gary-sweett, @gpx1000, @JoseEmilio-ARM, @r-potter, @sfricke-samsung, @wasimabbas-arm, @ZandroFargnoli

Hi reviewers, we're about to go live with the SDK and blog for Sync2 (targeting Monday).  Would be cool if we could include Sascha's sample at the same time.  Any chance you can break free to check this out?  Thanks!

mcj

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+			VkDependencyInfoKHR dependency_info{};
+			dependency_info.sType                    = VK_STRUCTURE_TYPE_DEPENDENCY_INFO_KHR;
+			dependency_info.bufferMemoryBarrierCount = 1;
+			dependency_info.pBufferMemoryBarriers    = &buffer_barrier;
+			vkCmdPipelineBarrier2KHR(draw_cmd_buffers[i], &dependency_info);
+		}
+
+		// Draw the particle system using the update vertex buffer
+		vkCmdBeginRenderPass(draw_cmd_buffers[i], &render_pass_begin_info, VK_SUBPASS_CONTENTS_INLINE);
+		VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:138]: I would suggest to prefer C++ style casting over C style.  Just a small nitpick.
`VkViewport viewport = vkb::initializers::viewport(static_cast<float>(width), static_cast<float>(height), 0.0f, 1.0f);`

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+
+	num_particles = static_cast<uint32_t>(attractors.size()) * PARTICLES_PER_ATTRACTOR;
+
+	// Initial particle positions
+	std::vector<Particle> particle_buffer(num_particles);
+
+	std::default_random_engine      rnd_engine(is_benchmark_mode() ? 0 : (unsigned) time(nullptr));
+	std::normal_distribution<float> rnd_distribution(0.0f, 1.0f);
+
+	for (uint32_t i = 0; i < static_cast<uint32_t>(attractors.size()); i++)
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:271]: FWIW, this loop has what looks like several transitory calculations that could be brought out of the loops as an optimization.  It's probably better to leave as is for readability.

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+ */
+
+/*
+ * Doing compute and graphics queue synchronization using VK_KHR_synchronization2
+ * Synchronization examples using this extension can be found at https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples
+ */
+
+#include "synchronization_2.h"
+
+Synchronization2::Synchronization2()
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:25]: constructor doesn't initialize num_particles

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+		}
+
+		// Draw the particle system using the update vertex buffer
+		vkCmdBeginRenderPass(draw_cmd_buffers[i], &render_pass_begin_info, VK_SUBPASS_CONTENTS_INLINE);
+		VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
+		vkCmdSetViewport(draw_cmd_buffers[i], 0, 1, &viewport);
+		VkRect2D scissor = vkb::initializers::rect2D(width, height, 0, 0);
+		vkCmdSetScissor(draw_cmd_buffers[i], 0, 1, &scissor);
+		vkCmdBindPipeline(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, graphics.pipeline);
+		vkCmdBindDescriptorSets(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, graphics.pipeline_layout, 0, 1, &graphics.descriptor_set, 0, NULL);
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:143]: More nitpicking: NULL -> nullptr for C++

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 5]
```diff
@@ -0,0 +1,866 @@
...
+		dependency_info.sType                    = VK_STRUCTURE_TYPE_DEPENDENCY_INFO_KHR;
+		dependency_info.bufferMemoryBarrierCount = 1;
+		dependency_info.pBufferMemoryBarriers    = &buffer_barrier;
+		vkCmdPipelineBarrier2KHR(compute.command_buffer, &dependency_info);
+	}
+
+	// First pass: Calculate particle movement
+	// -------------------------------------------------------------------------------------------------------
+	vkCmdBindPipeline(compute.command_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, compute.pipeline_calculate);
+	vkCmdBindDescriptorSets(compute.command_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, compute.pipeline_layout, 0, 1, &compute.descriptor_set, 0, 0);
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:203]: The final 0 should be a nullptr; while it's effectively the same, in a sample it's best to show best practices and be precise.

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+
+	// Second pass: Integrate particles
+	// -------------------------------------------------------------------------------------------------------
+	vkCmdBindPipeline(compute.command_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, compute.pipeline_integrate);
+	vkCmdDispatch(compute.command_buffer, num_particles / work_group_size, 1, 1);
+
+	// Release
+	if (graphics.queue_family_index != compute.queue_family_index)
+	{
+		VkBufferMemoryBarrier2KHR buffer_barrier{};
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:231]: declaration shadows a previous variable on line 203

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 7]
```diff
@@ -0,0 +1,866 @@
...
+		buffer_barrier.sType               = VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER_2_KHR;
+		buffer_barrier.srcAccessMask       = VK_ACCESS_2_SHADER_WRITE_BIT_KHR;
+		buffer_barrier.dstAccessMask       = VK_ACCESS_2_NONE_KHR;
+		buffer_barrier.srcStageMask        = VK_PIPELINE_STAGE_2_COMPUTE_SHADER_BIT_KHR;
+		buffer_barrier.dstStageMask        = VK_PIPELINE_STAGE_2_VERTEX_INPUT_BIT_KHR;
+		buffer_barrier.srcQueueFamilyIndex = compute.queue_family_index;
+		buffer_barrier.dstQueueFamilyIndex = graphics.queue_family_index;
+		buffer_barrier.buffer              = compute.storage_buffer->get_handle();
+		buffer_barrier.size                = compute.storage_buffer->get_size();
+		VkDependencyInfoKHR dependency_info{};
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:241]: dependency_info shadows a previous definition on line 217

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 8]
```diff
@@ -0,0 +1,866 @@
...
+	VkDescriptorBufferInfo            buffer_descriptor         = create_descriptor(*graphics.uniform_buffer);
+	VkDescriptorImageInfo             particle_image_descriptor = create_descriptor(textures.particle);
+	VkDescriptorImageInfo             gradient_image_descriptor = create_descriptor(textures.gradient);
+	std::vector<VkWriteDescriptorSet> write_descriptor_sets;
+	write_descriptor_sets = {
+	    vkb::initializers::write_descriptor_set(graphics.descriptor_set, VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, 0, &particle_image_descriptor),
+	    vkb::initializers::write_descriptor_set(graphics.descriptor_set, VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, 1, &gradient_image_descriptor),
+	    vkb::initializers::write_descriptor_set(graphics.descriptor_set, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 2, &buffer_descriptor),
+	};
+	vkUpdateDescriptorSets(get_device().get_handle(), static_cast<uint32_t>(write_descriptor_sets.size()), write_descriptor_sets.data(), 0, NULL);
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:406]: NULL -> nullptr

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 9]
```diff
@@ -0,0 +1,866 @@
...
+	    VK_DYNAMIC_STATE_SCISSOR};
+	VkPipelineDynamicStateCreateInfo dynamicState =
+	    vkb::initializers::pipeline_dynamic_state_create_info(
+	        dynamic_state_enables.data(),
+	        static_cast<uint32_t>(dynamic_state_enables.size()),
+	        0);
+
+	// Rendering pipeline
+	// Load shaders
+	std::array<VkPipelineShaderStageCreateInfo, 2> shader_stages;
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:459]: uninitialized variable.

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 10]
```diff
@@ -0,0 +1,866 @@
...
+
+	// Set some shader parameters via specialization constants
+	struct SpecializationData
+	{
+		uint32_t workgroup_size;
+		uint32_t shared_data_size;
+		float    gravity;
+		float    power;
+		float    soften;
+	} specialization_data;
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:597]: uninitialized (put a {} on 597)
`} specialization_data{};`

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 11]
```diff
@@ -0,0 +1,866 @@
...
+	            0,
+	            &storage_buffer_descriptor),
+	        // Binding 1 : Uniform buffer
+	        vkb::initializers::write_descriptor_set(
+	            compute.descriptor_set,
+	            VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
+	            1,
+	            &uniform_buffer_descriptor)};
+
+	vkUpdateDescriptorSets(get_device().get_handle(), static_cast<uint32_t>(compute_write_descriptor_sets.size()), compute_write_descriptor_sets.data(), 0, NULL);
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:581]: NULL -> nullptr

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 12]
```diff
@@ -0,0 +1,866 @@
...
+	build_compute_command_buffer();
+
+	// If necessary, acquire and immediately release the storage buffer, so that the initial acquire
+	// from the graphics command buffers are matched up properly.
+	if (graphics.queue_family_index != compute.queue_family_index)
+	{
+		VkCommandBuffer transfer_command;
+
+		// Create a transient command buffer for setting up the initial buffer transfer state
+		VkCommandBufferAllocateInfo command_buffer_allocate_info =
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:666]: command_buffer_allocate_info shadows a previous variable name on line 637.

File: samples/extensions/synchronization_2/synchronization_2.cpp [hunk 13]
```diff
@@ -0,0 +1,866 @@
...
+		VkDependencyInfoKHR dependency_info{};
+		dependency_info.sType                    = VK_STRUCTURE_TYPE_DEPENDENCY_INFO_KHR;
+		dependency_info.bufferMemoryBarrierCount = 2;
+		dependency_info.pBufferMemoryBarriers    = buffer_barriers;
+		vkCmdPipelineBarrier2KHR(transfer_command, &dependency_info);
+
+		VK_CHECK(vkEndCommandBuffer(transfer_command));
+
+		// Submit compute commands
+		VkSubmitInfo submit_info{};
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:709]: variable name shadows a previously defined one on line 650

File: samples/extensions/synchronization_2/synchronization_2.h [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+#else
+#	define PARTICLES_PER_ATTRACTOR 4 * 1024
+#endif
+
+class Synchronization2 : public ApiVulkanSample
+{
+  public:
+	uint32_t num_particles;
+	uint32_t work_group_size  = 128;
+	uint32_t shared_data_size = 1024;
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:39]: I personally prefer to use initializer lists for easier reference (everything is in the cpp file for the construction; but I realize this is a personal preference; num_particles isn't inited).

File: samples/extensions/synchronization_2/synchronization_2.h [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+	// SSBO particle declaration
+	struct Particle
+	{
+		glm::vec4 pos;        // xyz = position, w = mass
+		glm::vec4 vel;        // xyz = velocity, w = gradient texture position
+	};
+
+	Synchronization2();
+	~Synchronization2();
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:99]: virtual destructor.  mark with override.

File: samples/extensions/synchronization_2/synchronization_2.h [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+	virtual void request_gpu_features(vkb::PhysicalDevice &gpu) override;
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:100]: remove virtual keep override

File: samples/extensions/synchronization_2/synchronization_2.h [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+	void         setup_descriptor_set();
+	void         prepare_pipelines();
+	void         prepare_graphics();
+	void         prepare_compute();
+	void         prepare_uniform_buffers();
+	void         update_compute_uniform_buffers(float delta_time);
+	void         update_graphics_uniform_buffers();
+	void         draw();
+	bool         prepare(vkb::Platform &platform) override;
+	virtual void render(float delta_time) override;
```
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:116]: remove virtual keep override

File: shaders/synchronization_2/particle.vert [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+	const float spriteSize = 0.005 * inPos.w; // Point size influenced by mass (stored in inPos.w);
+
+	vec4 eyePos = ubo.modelview * vec4(inPos.x, inPos.y, inPos.z, 1.0); 
+	vec4 projectedCorner = ubo.projection * vec4(0.5 * spriteSize, 0.5 * spriteSize, eyePos.z, eyePos.w);
+	gl_PointSize = clamp(ubo.screendim.x * projectedCorner.x / projectedCorner.w, 1.0, 128.0);
+	
+	gl_Position = ubo.projection * eyePos;
+
+	outGradientPos = inVel.w;
+}
```
Comment (gpx1000) [inline shaders/synchronization_2/particle.vert:42]: add a newline at end of file

File: shaders/synchronization_2/particle_calculate.comp [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+		barrier();
+	}
+
+	particles[index].vel.xyz += ubo.deltaT * TIME_FACTOR * acceleration.xyz;
+
+	// Gradient texture position
+	particles[index].vel.w += 0.1 * TIME_FACTOR * ubo.deltaT;
+	if (particles[index].vel.w > 1.0)
+		particles[index].vel.w -= 1.0;
+}
```
Comment (gpx1000) [inline shaders/synchronization_2/particle_calculate.comp:89]: add newline at end of file

File: shaders/synchronization_2/particle_integrate.comp [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+#define TIME_FACTOR 0.05
+
+void main() 
+{
+	int index = int(gl_GlobalInvocationID);
+	vec4 position = particles[index].pos;
+	vec4 velocity = particles[index].vel;
+	position += ubo.deltaT * TIME_FACTOR * velocity;
+	particles[index].pos = position;
+}
```
Comment (gpx1000) [inline shaders/synchronization_2/particle_integrate.comp:48]: add newline

**Review by gpx1000 — DISMISSED** [review]
Everything looks good.  it worked well for me in Linux.
I added some minor requests in the code.  None of them are worth holding back an approval.  Fixing them are optional.

**Comment by SaschaWillems** [comment]
Thanks a lot for the exhaustive code comments @gpx1000. They all do make sense, and also apply to the compute N-Body sample this is based on. Once this sample has been merged, I'll do another PR for your suggestions.

**Review by carlosadcq — DISMISSED** [review]

File: samples/extensions/synchronization_2/synchronization_2.h [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+	virtual void resize(const uint32_t width, const uint32_t height) override;
```
Comment (mklefrancois) [inline samples/extensions/synchronization_2/synchronization_2.h:117]: Remove redundant 'virtual'

File: samples/extensions/synchronization_2/synchronization_2.cpp (hunk 9 above)
Comment (mklefrancois) [reply samples/extensions/synchronization_2/synchronization_2.cpp:459]: I think this is fine for readability. The next two lines are initializing `shader_stages`

File: samples/extensions/synchronization_2/synchronization_2.cpp (hunk 10 above)
Comment (mklefrancois) [reply samples/extensions/synchronization_2/synchronization_2.cpp:597]: And to initialize the struct with the same values as in `particle_calulate.comp`
```
        struct SpecializationData
	{
		uint32_t workgroup_size{0};
		uint32_t shared_data_size{1024};
		float    gravity{0.002f};
		float    power{0.75f};
		float    soften{0.05f};
	} specialization_data;
```

**Review by mklefrancois — DISMISSED** [review]
Work and tested with
- NVIDIA A6000
- Windows Driver 457.96
- Vulkan 1.2.170.0

**Pull request merged**